### DISCUSSION TITLE
High performance queries with thousands of orders

### DISCUSSION THREAD

**Discussion by ESPONI** [woocommerce/woocommerce#35921] [WooCommerce HPOS]
Hi!

Trying to avoid TL;DR :)
 
* We sell physical products and tickets.
* A ticket is “simply” a product that has a meta in order_item: is_ticket=1

Everything is ok until the real scenario:
* 150.000 ~ 160.000 orders/year.
* Many orders has more than 1 line item, may be 5…
* Some are products, 95% are tickets
* Some tickets has another meta: ‘is_invitation=1’

Question:
How to perform thouse high performance queries using woocommerce classes and helpers, in a milisecond, like we do in SQL directly.
(We don’t want to lose compatibility, hardcoding SQL)

Basically, count total quantity sold of…
* a given product_id
* total invitations
* and similar to those…
 
Best regards!!!

**Comment by vedanshujain** [comment]
Probably not the answer you are looking for, but the fastest way for highly customized cases like these, you would want to write a custom SQL query. Alternatively, for the functionality missing in WC_Query class that prevents you from doing this, you can consider raising an issue so that we can support it (or even better, send a PR). The default meta queries that currently exist might be slow, they will at least do a join with the main table.

**Comment by ESPONI** [comment]
Hi! Thanks vedanshujain.

would be awesome a set of helpers specialized in counting.

**Comment by lsinger** [comment]
> would be awesome a set of helpers specialized in counting

@ESPONI could you help us understand your requirements a bit better? If you could design your dream API, what method(s) would you call on which objects, which arguments would they take and what would they return?
